packages/date-time/src/components/Calendar/CalendarYear/CalendarYear.base.tsx

```diff
@@ -0,0 +1,458 @@
+        {this._onRenderNav()}
+      </div>
+    );
+  }
+  private _onRenderTitle = () => {
+    if (this.props.onRenderTitle) {
+      return this.props.onRenderTitle(this.props);
+    }
+    return <CalendarYearTitle {...this.props} />;
+  };
+  private _onRenderNav = () => {
+    return <CalendarYearNav {...this.props} />;
+  };
+}
+export interface ICalendarYearState {
+  fromYear: number;
+  navigatedYear?: number;
+  selectedYear?: number;
+}
+
+export class CalendarYearBase extends BaseComponent<ICalendarYearProps, ICalendarYearState> implements ICalendarYear {
+  private _gridRef: CalendarYearGrid;
+
+  constructor(props: ICalendarYearProps) {
+    super(props);
+    this.state = this._getState(this.props);
+  }
+
+  public componentWillReceiveProps(nextProps: ICalendarYearProps): void {
+    this.setState(this._getState(nextProps));
+  }
+
+  public focus(): void {
+    if (this._gridRef) {
+      this._gridRef.focus();
+    }
+  }
+
+  public render(): JSX.Element {
+    const { styles, theme, className } = this.props;
+
+    const classNames = getClassNames(styles, {
+      theme: theme!,
+      className: className
+    });
+
+    return (
+      <div className={css('ms-DatePicker-yearPicker', classNames.root)}>
+        {this._renderHeader()}
+        {this._renderGrid()}
+      </div>
+    );
+  }
+
+  private _onNavNext = () => {
+    this.setState({ fromYear: this.state.fromYear + CELL_COUNT });
+  };
+
+  private _onNavPrev = () => {
+    this.setState({ fromYear: this.state.fromYear - CELL_COUNT });
+  };
+
+  private _renderHeader = (): React.ReactNode => {
+    return (
+      <CalendarYearHeader
+        {...this.props}
+        fromYear={this.state.fromYear}
+        toYear={this.state.fromYear + CELL_COUNT - 1}
+        onSelectPrev={this._onNavPrev}
+        onSelectNext={this._onNavNext}
+      />
+    );
+  };
+
+  private _renderGrid = (): React.ReactNode => {
+    return (
+      <CalendarYearGrid
+        {...this.props}
+        fromYear={this.state.fromYear}
+        toYear={this.state.fromYear + CELL_COUNT - 1}
+        ref={this._onGridRef}
+      />
+    );
+  };
+
+  private _onGridRef = (ref: CalendarYearGrid) => {
+    this._gridRef = ref;
+  };
+
+  private _getState(props: ICalendarYearProps): ICalendarYearState {
+    const { selectedYear, navigatedYear } = props;
+    const rangeYear = selectedYear || navigatedYear || new Date().getFullYear();
+    const fromYear = Math.floor(rangeYear / 10) * 10;
+    return {
+      fromYear: fromYear,
+      navigatedYear: navigatedYear,
+      selectedYear: selectedYear
+    };
+  }
+}
+
+interface ICalendarYearGridCellProps extends ICalendarYearProps {
+  year: number;
+  current?: boolean;
+  selected?: boolean;
+  disabled?: boolean;
+  onSelectYear?: (year: number) => void;
+  onRenderYear?: (year: number) => React.ReactNode;
+}
+
+interface ICalendarYearGridProps extends ICalendarYearProps, ICalendarYearRange {
+  selectedYear?: number;
+}
```

Comment:
Is there a reason these interfaces are at the bottom of the file?   There already one at the top of the file

Not sure if there is a hard and fast rule about this, but I typically see them at the top.

Reply:
Direct copy from other component. Will move to top.